### DISCUSSION TITLE
What type should I use for generated schema?

### DISCUSSION THREAD

**Discussion by frederikhors** [urql-graphql/urql#1900] [Q&A] [answered]
I'm generating schema for awareness using codegen like this:

- codegen.yml:

```yml
overwrite: true
schema: ./schema/*.graphql
generates:
  ./schema.json:
    plugins:
      - urql-introspection
```

it works and generates the same json as if I used the code [explained on docs](https://formidable.com/open-source/urql/docs/graphcache/schema-awareness/#installation--setup).

### THE QUESTION

When I `import` schema in my `urql-init.ts` file what type should I use? What syntax?

If I simpy use this:

```ts
import schema from "./schema.json";

const cache = cacheExchange({ schema });
```

Typescript complains with:

```
...
Property 'possibleTypes' is missing in type '{ kind: string; name: string; fields: ({ name: string; type: { kind: string; ofType: { kind: string; name: string; ofType?: undefined; }; name?: undefined; }; args: any[]; } | { name: string; type: { kind: string; name: string; ofType?: undefined; }; args: any[]; } | { ...; } | { ...; } | { ...; })[]; interfaces: an...' but required in type 'IntrospectionInterfaceType'.ts(2322)

getIntrospectionQuery.d.ts(90, 12): 'possibleTypes' is declared here.

types.d.ts(96, 5): The expected type comes from property 'schema' which is declared here on type 'Partial<CacheExchangeOpts>'
```

so I can cast schema with which type?

Is it ok to use the below?

```ts
import type { IntrospectionData } from "@urql/exchange-graphcache/dist/types/ast";

import rawSchema from "./schema.json";

const schema = rawSchema as IntrospectionData;

const cache = cacheExchange({ schema });
```

**Is there a better way?**

Thanks!

_As soon as you answer me I will PR for update [docs](https://formidable.com/open-source/urql/docs/graphcache/schema-awareness/#integrating-a-schema)._

**Comment by JoviDeCroock** [comment, marked as answer]
I think the cast is a good way, generally the Json type inference doesn't work too well
